### DISCUSSION TITLE
What's the recommended way to implement authentication?

### DISCUSSION THREAD

**Discussion by hfoletto** [nuxt-community/apollo-module#471] [Q&A]
Hello everyone, greetings!

I'm a long time Vue developer and recently decided to start playing with GraphQL.
I'm using Nuxt 3, so this module is very useful!

In past projects I was using Nuxt 2 with a standard REST API. For authentication I used the nuxt auth module: https://auth.nuxtjs.org/
After installing, the usage was very straight-forward:
```
this.$auth.loginWith('schemeName', {
  data: {
    email: '',
    password: ''
  }
})
```
I would use the above code in the login page, and I'm done. I'd have access to many properties and methods across the whole application: https://auth.nuxtjs.org/api/auth

Now with Apollo, I'm a bit lost on how to handle it. In the backend I already have the login, logout mutations, however I think some sort of adapter or just a convention is what I'm looking for.

Thank you!

**Comment by christohill** [comment]
@hfoletto Have you made any headway with this? I'm in a similar scenario and not able to find much documentation on it.

**Comment by hfoletto** [comment]
Hello @christohill,

I didn't find any kind of adapter or convention for that, so I decided to do it my own.
For the most part, I tried to mimic how the auth package worked, and so far it's actually working great for me.

Assuming you already have graphql login/logout/register mutations on your API, this is what I did:

I created a `auth.ts` pinia store, with the following contents:
```
import { defineStore } from 'pinia'

type GetMe = {
  me: {
    email: string,
    name: string
  }
}

export const useAuthStore = defineStore('user', {
  state: () => {
    return {
      user: null as {
        name: string,
        email: string
      } | null,
    }
  },
  getters: {
    loggedIn: state => state.user !== null,
  },
  actions: {
    async fetchUser () {
      const query = gql`
        query getMe {
          me {
            email,
            name
          }
        }
      `
      const { data, error } = await useAsyncQuery<GetMe>(query)
      if (!error.value && data.value?.me)
        this.user = data.value?.me
    },
  },
})
```
I also created a plugin to fetchUser on initial page load:
`/plugins/auth.ts`
```
import { useAuthStore } from '@/store/auth'
export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.hook('app:beforeMount', async () => {
    const store = useAuthStore()
    await store.fetchUser()
  })
})
```

I could have added the login and logout actions in the store as well, but decided to have those in the login page and default layout, respectively:

```
const loginMutationQuery = gql`
  mutation login($email: String!, $password: String!) {
    login (email: $email, password: $password) {
      name,
      email
    }
  }
`

const { mutate: loginMutation, error, loading } = useMutation(loginMutationQuery)

async function login () {
  const { data } = await loginMutation({ email: email.value, password: password.value })
  if (!error.value) {
    store.$patch({
      user: { ...data.login },
    })
    $toast.show({
      type: 'success',
      message: 'Logged in successfully',
    })
    window.setTimeout(() => {
      return navigateTo('/')
    }, 1000)
  }
}
```

```
const logoutMutationQuery = gql`
  mutation logout {
    logout {
      name
    }
  }
`

const { mutate: logoutMutation, error, loading } = useMutation(logoutMutationQuery)

async function logout () {
  await logoutMutation()
  if (!error.value) {
    store.$patch({
      user: null,
    })
    $toast.show({
      type: 'info',
      message: 'Logged out successfully',
    })
    window.setTimeout(() => {
      return navigateTo('/')
    }, 1000)
  }
}
```

This is working great to me, but I'm not sure if that's the correct way of doing it.

I'm considering moving the logout function to the store actions. The login I still prefer to have in the login page since it makes error handling easier (credentials invalid and etc).
